### ISSUE TITLE
`ccomplete.vim` can not handle array cases very well

### ISSUE THREAD

**Issue by xndcn** [vim/vim#3349]
Try these cases, only array case 2'' and case 7'' will pass.
It seem that `ccomplete.vim` can not handle array cases very well.

```c
typedef struct Foo {
    typedef struct Bar {
        int foobar;
    } Bar;
    Bar bar;
    Bar bar_array[0];

    Bar *bar_pointer;
    Bar *bar_pointer_array[0];
    int data;
} Foo;

Foo global_foo;
Foo global_foo_array[0];

int main() {
    Foo  foo;
    Foo *foo_array[0];

    // without ./->
    foo                                        // case 1
    foo.bar                                    // case 2
    global_foo                                 // case 3
    extern_foo                                 // case 4
    extern_foo->bar_pointer                    // case 5

    // with ./->
    foo.                                       // case 6
    foo.bar.                                   // case 7
    global_foo.bar.                            // case 8
    extern_foo->bar_pointer->                  // case 9

    // array
    foo_array[0]                               // case 1'
    foo_array[0]->bar_array[0]                 // case 2'
    foo.bar_array[0]                           // case 2''
    global_foo_array[0]                        // case 3'
    extern_foo_array[0]                        // case 4'
    extern_foo_array[0].bar_pointer_array[0]   // case 5'

    foo_array[0]->                             // case 6'
    foo_array[0]->bar_array[0].                // case 7'
    foo.bar_array[0].                          // case 7''
    global_foo_array[0].bar_array[0].          // case 8'
    extern_foo_array[0]->bar_pointer_array[0]. // case 9'

    return 0;
}
```

**Comment by xndcn** [comment]
I'm trying to fix it.
I think `items` with `[...]` should be treated specially:
1. If `items` like `aa[...][...]...[...]`, then it's a local/global array variable
2. for `items` like `aa[...].bb`, `[...]` should be skip over in function `s:StructMembers` when we try to match members of `aa`

**Comment by brammool** [comment]
> Try these cases, only array case 2'' and case 7'' will pass.
> It seem that `ccomplete.vim` can not handle array cases very well.

I haven't had time to make this work better.  There is also
cppcomplete.vim, but I don't think it was maintained either.
There are probably better scripts available somewhere.


-- 
./configure
Checking whether build environment is sane ...
build environment is grinning and holding a spatula.  Guess not.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by xndcn** [comment]
@brammool Got it, thanks!
Sometimes I think the pre-installed ccomplete.vim is good enough to handle most of work.
Otherwise I will consider more heavy auto-complete plugin.

**Comment by chrisbra** [comment]
@xndcn So are you actually proposing a change/patch here?

**Comment by xndcn** [comment]
@chrisbra Emm... Yes, please review this change if it can help, thanks.
https://github.com/vim/vim/pull/3350

**Comment by chrisbra** [comment]
fixed by 20aac6c1126988339611576d425965a25a7776
